File: packages/php/src/Uncaught.php
<?php

declare(strict_types=1);

namespace Uncaught;

/**
 * Static entry-point for the Uncaught SDK.
 *
 * Usage:
 *   Uncaught::init(['environment' => 'production', 'release' => '1.0.0']);
 *   Uncaught::captureException($e);
 */
final class Uncaught
{
    private static ?Client $client = null;

    /**
     * Initialise the Uncaught SDK. Calling this more than once replaces
     * the previous client instance.
     *
     * @param array<string, mixed> $config
     */
    public static function init(array $config = []): Client
    {
        self::$client = new Client($config);

        // Register as the global error / exception handler
        set_exception_handler([self::class, 'handleException']);
        set_error_handler([self::class, 'handleError']);
        register_shutdown_function([self::class, 'handleShutdown']);

        return self::$client;
    }

    /**
     * Return the current singleton client, or null if init() has not been called.
     */
    public static function getClient(): ?Client
    {
        return self::$client;
    }

    /**
     * Capture an exception.
     */
    public static function captureException(\Throwable $e, array $context = []): ?string
    {
        return self::$client?->captureException($e, $context);
    }

    /**
     * Capture a plain message.
     */
    public static function captureMessage(string $message, string $level = 'info'): ?string
    {
        return self::$client?->captureMessage($message, $level);
    }

    /**
     * Add a breadcrumb.
     */
    public static function addBreadcrumb(
        string $type,
        string $category,
        string $message,
        ?array $data = null,
        ?string $level = null
    ): void {
        self::$client?->addBreadcrumb($type, $category, $message, $data, $level);
    }

    /**
     * Set user context.
     */
    public static function setUser(?array $user): void
    {
        self::$client?->setUser($user);
    }

    /**
     * Global exception handler.
     */
    public static function handleException(\Throwable $e): void
    {
        self::$client?->captureException($e, ['level' => 'fatal']);
    }

    /**
     * Global error handler.
     */
    public static function handleError(
        int $errno,
        string $errstr,
        string $errfile = '',
        int $errline = 0
    ): bool {
        if (!(error_reporting() & $errno)) {
            return false;
        }

        $exception = new \ErrorException($errstr, 0, $errno, $errfile, $errline);
        $level = match (true) {
            ($errno & (E_ERROR | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR)) !== 0 => 'fatal',
            ($errno & (E_WARNING | E_CORE_WARNING | E_COMPILE_WARNING | E_USER_WARNING)) !== 0 => 'warning',
            default => 'error',
        };

        self::$client?->captureException($exception, ['level' => $level]);
        return false;
    }

    /**
     * Shutdown handler for fatal errors.
     */
    public static function handleShutdown(): void
    {
        $error = error_get_last();
        if ($error !== null && in_array($error['type'], [E_ERROR, E_CORE_ERROR, E_COMPILE_ERROR, E_PARSE], true)) {
            $exception = new \ErrorException(
                $error['message'],
                0,
                $error['type'],
                $error['file'],
                $error['line']
            );
            self::$client?->captureException($exception, ['level' => 'fatal']);
        }
    }

    private function __construct()
    {
    }
}

File: packages/go/uncaught.go
// ---------------------------------------------------------------------------
// uncaught-go — Public API
// ---------------------------------------------------------------------------
//
// Package uncaught provides error monitoring for Go applications with
// automatic fingerprinting, breadcrumb tracking, and AI-powered fix prompts.
//
// Quick start:
//
//	client := uncaught.Init(&uncaught.Config{
//	    Environment: "production",
//	    Release:     "v1.0.0",
//	})
//
//	// Capture errors
//	client.CaptureError(err, nil)
//
//	// Add breadcrumbs for context
//	client.AddBreadcrumb(uncaught.Breadcrumb{
//	    Type:     uncaught.BreadcrumbNavigation,
//	    Category: "http",
//	    Message:  "GET /api/users",
//	})
// ---------------------------------------------------------------------------

package uncaught

import "sync"

var (
	globalMu     sync.RWMutex
	globalClient *UncaughtClient
)

// Init initialises the Uncaught SDK with the given configuration.
// Calling this more than once replaces the previous client instance.
// Returns the new client.
func Init(config *Config) *UncaughtClient {
	client := NewClient(config)

	globalMu.Lock()
	globalClient = client
	globalMu.Unlock()

	return client
}

// GetClient returns the current global singleton client, or nil if Init()
// has not been called.
func GetClient() *UncaughtClient {
	globalMu.RLock()
	defer globalMu.RUnlock()
	return globalClient
}

// CaptureError is a convenience wrapper that captures an error using the
// global client. Returns empty string if the global client is not initialised.
func CaptureError(err interface{}, ctx *CaptureContext) string {
	client := GetClient()
	if client == nil {
		return ""
	}
	return client.CaptureError(err, ctx)
}

// CaptureMessage is a convenience wrapper that captures a message using the
// global client.
func CaptureMessage(message string, level SeverityLevel) string {
	client := GetClient()
	if client == nil {
		return ""
	}
	return client.CaptureMessage(message, level)
}

// AddBreadcrumb is a convenience wrapper that adds a breadcrumb using the
// global client.
func AddBreadcrumb(crumb Breadcrumb) {
	client := GetClient()
	if client == nil {
		return
	}
	client.AddBreadcrumb(crumb)
}

// SetUser is a convenience wrapper that sets the user context on the global client.
func SetUser(user *UserInfo) {
	client := GetClient()
	if client == nil {
		return
	}
	client.SetUser(user)
}

// Flush flushes all queued events on the global client.
func Flush() error {
	client := GetClient()
	if client == nil {
		return nil
	}
	return client.Flush()
}

// Recover is a convenience function to be used in a defer statement.
// It recovers from panics and captures the error using the global client.
//
//	func handler() {
//	    defer uncaught.Recover(nil)
//	    // ... code that might panic
//	}
func Recover(ctx *CaptureContext) {
	client := GetClient()
	if client == nil {
		return
	}
	client.Recover(ctx)
}
